src/wzapi.cpp

```diff
@@ -0,0 +1,4339 @@
+	// check against building in a gateway, as this can seriously block AI passages
+	for (auto psGate : gwGetGateways())
+	{
+		for (x = xx; x <= xBR; x++)
+		{
+			for (y = yy; y <= yBR; y++)
+			{
+				if ((x >= psGate->x1 && x <= psGate->x2) && (y >= psGate->y1 && y <= psGate->y2))
+				{
+					return false;
+				}
+			}
+		}
+	}
+
+	// can you get past it?
+	y = yTL;	// top
+	for (x = xTL; x != xBR + 1; x++)
+	{
+		if (fpathBlockingTile(x, y, PROPULSION_TYPE_WHEELED))
+		{
+			count++;
+			break;
+		}
+	}
+
+	y = yBR;	// bottom
+	for (x = xTL; x != xBR + 1; x++)
+	{
+		if (fpathBlockingTile(x, y, PROPULSION_TYPE_WHEELED))
+		{
+			count++;
+			break;
+		}
+	}
+
+	x = xTL;	// left
+	for (y = yTL + 1; y != yBR; y++)
+	{
+		if (fpathBlockingTile(x, y, PROPULSION_TYPE_WHEELED))
+		{
+			count++;
+			break;
+		}
+	}
+
+	x = xBR;	// right
+	for (y = yTL + 1; y != yBR; y++)
+	{
+		if (fpathBlockingTile(x, y, PROPULSION_TYPE_WHEELED))
+		{
+			count++;
+			break;
+		}
+	}
+
+	//make sure this location is not blocked from too many sides
+	if ((count <= maxBlockingTiles) || (maxBlockingTiles == -1))
+	{
+		return true;
+	}
+	return false;
+}
+
+//-- ## pickStructLocation(droid, structure type, x, y[, maxBlockingTiles])
+//--
+//-- Pick a location for constructing a certain type of building near some given position.
+//-- Returns an object containing "type" POSITION, and "x" and "y" values, if successful.
+//--
+optional<scr_position> wzapi::pickStructLocation(WZAPI_PARAMS(const DROID *psDroid, std::string statName, int startX, int startY, optional<int> _maxBlockingTiles))
+{
+	SCRIPT_ASSERT({}, context, psDroid, "No valid droid provided");
+	const int player = psDroid->player;
+	SCRIPT_ASSERT_PLAYER({}, context, player);
+	int index = getStructStatFromName(WzString::fromUtf8(statName));
+	SCRIPT_ASSERT({}, context, index >= 0, "%s not found", statName.c_str());
+	STRUCTURE_STATS	*psStat = &asStructureStats[index];
+	SCRIPT_ASSERT({}, context, psStat, "No such stat found: %s", statName.c_str());
+
+	int numIterations = 30;
+	bool found = false;
+	int incX, incY, x, y;
+	int maxBlockingTiles = 0;
+
+	SCRIPT_ASSERT({}, context, startX >= 0 && startX < mapWidth && startY >= 0 && startY < mapHeight, "Bad position (%d, %d)", startX, startY);
+
+	if (_maxBlockingTiles.has_value())
+	{
+		maxBlockingTiles = _maxBlockingTiles.value();
+	}
+
+	x = startX;
+	y = startY;
+
+	Vector2i offset(psStat->baseWidth * (TILE_UNITS / 2), psStat->baseBreadth * (TILE_UNITS / 2));
+
+	// save a lot of typing... checks whether a position is valid
+#define LOC_OK(_x, _y) (tileOnMap(_x, _y) && \
+                        (!psDroid || fpathCheck(psDroid->pos, Vector3i(world_coord(_x), world_coord(_y), 0), PROPULSION_TYPE_WHEELED)) \
+                        && validLocation(psStat, world_coord(Vector2i(_x, _y)) + offset, 0, player, false) && structDoubleCheck(psStat, _x, _y, maxBlockingTiles))
+
+	// first try the original location
+	if (LOC_OK(startX, startY))
+	{
+		found = true;
```

Comment:
a `goto endstructloc;` could go here in  `pickStructLocation` for a small optimization. Or what that return statement has there.